packages/api/src/utils/types.ts

```diff
@@ -1,253 +1,160 @@
+    onlySupport?: WireFormat;
   };
-}
 
 /**
- * SSZ factory helper + typed to return responses of type `{data: T; executionOptimistic: boolean}`
+ * Handles translation between Endpoint["args"] and Endpoint["request"]
  */
-export function ContainerDataExecutionOptimistic<T>(
-  dataType: TypeJson<T>
-): TypeJson<{data: T; executionOptimistic: boolean}> {
-  return {
-    toJson: ({data, executionOptimistic}) => ({
-      data: dataType.toJson(data),
-      execution_optimistic: executionOptimistic,
-    }),
-    fromJson: ({data, execution_optimistic}: {data: unknown; execution_optimistic: boolean}) => {
-      return {
-        data: dataType.fromJson(data),
-        executionOptimistic: execution_optimistic,
-      };
-    },
-  };
+export type RequestCodec<E extends Endpoint> = E["method"] extends "GET"
+  ? RequestWithoutBodyCodec<E>
+  : "body" extends keyof E["request"]
+    ? RequestWithBodyCodec<E>
+    : RequestWithoutBodyCodec<E>;
+
+export function isRequestWithoutBody<E extends Endpoint>(
+  definition: RouteDefinition<E>
+): definition is RouteDefinition<E> & {req: RequestWithoutBodyCodec<E>} {
+  return definition.method === "GET" || definition.req.schema.body === undefined;
 }
 
-/**
- * SSZ factory helper + typed to return responses of type
- * ```
- * data: T
- * version: ForkName
- * ```
- */
-export function WithVersion<T>(getType: (fork: ForkName) => TypeJson<T>): TypeJson<{data: T; version: ForkName}> {
-  return {
-    toJson: ({data, version}) => ({
-      data: getType(version ?? ForkName.phase0).toJson(data),
-      version,
-    }),
-    fromJson: ({data, version}: {data: unknown; version: string}) => {
-      // Teku returns fork as UPPERCASE
-      version = version.toLowerCase();
-
-      // Un-safe external data, validate version is known ForkName value
-      if (!(version in ForkName)) throw Error(`Invalid version ${version}`);
-
-      return {
-        data: getType(version as ForkName).fromJson(data),
-        version: version as ForkName,
-      };
-    },
-  };
-}
+// Response codec
 
-/**
- * SSZ factory helper to wrap an existing type with `{executionOptimistic: boolean}`
- */
-export function WithExecutionOptimistic<T extends {data: unknown}>(
-  type: TypeJson<T>
-): TypeJson<T & {executionOptimistic: boolean}> {
-  return {
-    toJson: ({executionOptimistic, ...data}) => ({
-      ...(type.toJson(data as unknown as T) as Record<string, unknown>),
-      execution_optimistic: executionOptimistic,
-    }),
-    fromJson: ({execution_optimistic, ...data}: T & {execution_optimistic: boolean}) => ({
-      ...type.fromJson(data),
-      executionOptimistic: execution_optimistic,
-    }),
-  };
-}
+export type ResponseDataCodec<T, M> = {
+  toJson: (data: T, meta: M) => unknown; // server
+  fromJson: (data: unknown, meta: M) => T; // client
+  serialize: (data: T, meta: M) => Uint8Array; // server
+  deserialize: (data: Uint8Array, meta: M) => T; // client
```

Comment:
I don't like adding such comments to code, if some code need comments forexplanation then better be add as doc comment than inline. 

```suggestion
  toJson: (data: T, meta: M) => unknown;
  fromJson: (data: unknown, meta: M) => T;
  serialize: (data: T, meta: M) => Uint8Array;
  deserialize: (data: Uint8Array, meta: M) => T;
```

Reply:
I am usually also not a fan of comments at the end of the line but here it is kinda fitting. Doesn't make sense to add this as jsdoc, and imo for me the methods are clear from where those are called (server vs. client) but for someone looking at the implementation the first time might be helpful, or at least safe a few seconds of brain power.

can we keep them? :pray: